lib/src/model.dart

```diff
@@ -4046,16 +4073,260 @@ class ModelFunctionTyped extends ModelElement
+  /// [input]. If the case type matches that of [caseDefault], then it
+  /// returns an empty string, so as not to add identifiers to strings that
+  /// already match the expected case type.
+  static String _getCaseSuffix(String input, CaseType caseDefault) {
+    CaseType caseType = _getCaseType(input);
+    if (caseDefault == caseType) return '';
+    return '-${enumToString(caseType)}';
+  }
+
+  /// Returns true if all characters in the given [value] are uppercase.
+  static bool _isAllUpper(String value) => value.toUpperCase() == value;
+
+  /// Returns true if all characters in the given [value] are lowercase.
+  static bool _isAllLower(String value) => value.toLowerCase() == value;
+
+  /// Returns the [CaseType] of the given [input]. The [input] cannot be
+  /// the empty string, or null.
+  static CaseType _getCaseType(String input) {
+    assert(input.isNotEmpty);
+    assert(input != null);
+    if (_isAllLower(input)) return CaseType.lower;
+    if (_isAllUpper(input)) return CaseType.upper;
+    int firstLetter = input.indexOf(new RegExp(r'[A-Za-z]'));
+    // firstLetter can't be -1: it would have returned true for isAllLower
+    // in that case.
+    if (_isAllUpper(input[firstLetter])) return CaseType.camel;
+    return CaseType.lowerCamel;
+  }
+
+  /// Returns a string that encodes the case "runs" of a string, which is
+  /// basically a run-length encoding of a bitfield containing the "case" of a
+  /// letter at that position in the string. This gives us a unique fingerprint
+  /// for the case information of a string.
+  static String _getCaseRuns(String input) {
+    assert(input.length <= 0xffff,
+        'Identifier strings longer than 65K not supported.');
+    if (input.isEmpty) return '';
+    if (input.length == 1) return '-1';
+    final int firstLetter = input.indexOf(new RegExp(r'[A-Za-z]'));
+    if (firstLetter == -1) return '';
+    List<int> runs = [];
+    int lastChange = 0;
+    CaseType lastCase = _getCaseType(input.substring(0, 1));
+    for (int i = firstLetter; i < input.length; ++i) {
+      CaseType thisCase = _getCaseType(input[i]);
+      if (thisCase != lastCase) {
+        lastCase = thisCase;
+        runs.add(lastChange);
+        lastChange = 0;
+      }
+      lastChange++;
+    }
+    runs.add(lastChange);
+    while (runs.length % 2 != 0) {
+      runs.insert(0, 0);
+    }
+    List<String> wordRuns = [];
+    for (var run in runs) {
+      if (input.length <= 0xf) {
+        wordRuns.add(run.toRadixString(16));
+      } else if (input.length <= 0xff) {
+        wordRuns.add(run.toRadixString(16).padLeft(2, '0'));
+      } else if (input.length > 0xff) {
+        wordRuns.add(run.toRadixString(16).padLeft(4, '0'));
+      }
+    }
+    return wordRuns.join('');
+  }
+
+  /// Builds a cache of unique names to speed up access to them once all the
+  /// names are collected. Only runs once, on the first access to a unique name
+  /// for this [NameRegistry].
+  void _buildCache() {
```

Comment:
nit:  typical dartdoc style is to build the cached item as part of the getter wrapping a variable.  For private members, this can look something like:

```dart
Thingy __theThing;
Thingy get _theThing {
  if (__theThing == null) {
    _theThing = new Thingy(doSomeStuff());
  }
  return _theThing;
}
```

Not super strict on it, but it would be nice for readability to make this consistent.

Reply:
I can't *quite* get there, because it needs to supply an argument that isn't always the `name`. (a unique `dirName` for libraries is also found using this function).